File: examples/elixir/lib/reverse.ex
defmodule Revaddr.Reverse do
  @moduledoc """
  Reverse-geocode a single lat/lon with the RevAddr HTTP API.

  ## Run

      export REVADDR_API_KEY=sk_live_...
      mix deps.get
      mix run -e "Revaddr.Reverse.main([])"
      mix run -e "Revaddr.Reverse.main([\\"37.7749\\", \\"-122.4194\\"])"

  Uses Erlang `:httpc` plus Jason. Suitable for Phoenix controllers or Oban jobs.
  """

  @default_base "https://api.revaddr.com"

  @doc """
  GET /v1/reverse and return the inner `"result"` map.
  """
  def reverse(lat, lon, api_key, base_url \\ @default_base)
      when is_number(lat) and is_number(lon) and is_binary(api_key) do
    base = String.trim_trailing(base_url, "/")
    # URI.encode_query builds lat/lon query string.
    query = URI.encode_query(%{"lat" => lat, "lon" => lon})
    url = String.to_charlist("#{base}/v1/reverse?#{query}")

    headers = [
      {~c"x-api-key", String.to_charlist(api_key)},
      {~c"Accept", ~c"application/json"}
    ]

    # inets httpc: ensure started in mix run / releases.
    :inets.start()
    :ssl.start()

    request = {url, headers}

    case :httpc.request(:get, request, [timeout: 30_000], body_format: :binary) do
      {:ok, {{_, status, _}, _headers, body}} when status >= 200 and status < 300 ->
        case Jason.decode(body) do
          {:ok, %{"result" => result}} when is_map(result) -> {:ok, result}
          {:ok, _} -> {:error, :missing_result}
          {:error, err} -> {:error, err}
        end

      {:ok, {{_, status, _}, _, body}} ->
        {:error, {:http, status, body}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc false
  def main(args) do
    api_key = System.get_env("REVADDR_API_KEY") |> to_string() |> String.trim()

    if api_key == "" do
      IO.puts(:stderr, "Set REVADDR_API_KEY to your RevAddr API key.")
      System.halt(1)
    end

    base =
      case System.get_env("REVADDR_BASE_URL") do
        nil -> @default_base
        "" -> @default_base
        v -> v
      end

    # Website hero default: White House.
    {lat, lon} =
      case args do
        [a, b | _] -> {String.to_float(a), String.to_float(b)}
        _ -> {38.8977, -77.0365}
      end

    case reverse(lat, lon, api_key, base) do
      {:ok, result} ->
        IO.puts(Map.get(result, "formatted_address") || "(no formatted_address)")
        IO.puts(Jason.encode!(result, pretty: true))

      {:error, reason} ->
        IO.puts(:stderr, inspect(reason))
        System.halt(1)
    end
  end
end

File: examples/php/reverse.php
<?php
/**
 * Reverse-geocode a single point with RevAddr (PHP 8+, cURL extension).
 *
 * Run:
 *   export REVADDR_API_KEY=sk_live_...
 *   php reverse.php
 *   php reverse.php 37.7749 -122.4194
 */

declare(strict_types=1);

const DEFAULT_BASE = 'https://api.revaddr.com';

$apiKey = getenv('REVADDR_API_KEY') ?: '';
if ($apiKey === '') {
    fwrite(STDERR, "Set REVADDR_API_KEY to your RevAddr API key.\n");
    exit(1);
}

$base = getenv('REVADDR_BASE_URL') ?: DEFAULT_BASE;
$base = rtrim($base, '/');

// Website hero default: White House.
$lat = isset($argv[1]) ? (float) $argv[1] : 38.8977;
$lon = isset($argv[2]) ? (float) $argv[2] : -77.0365;

// http_build_query encodes lat/lon correctly for the query string.
$query = http_build_query(['lat' => $lat, 'lon' => $lon], '', '&', PHP_QUERY_RFC3986);
$url = $base . '/v1/reverse?' . $query;

$ch = curl_init($url);
if ($ch === false) {
    fwrite(STDERR, "curl_init failed\n");
    exit(1);
}

// CURLOPT_HTTPHEADER carries the secret; never put keys in the URL.
curl_setopt_array($ch, [
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_TIMEOUT => 30,
    CURLOPT_HTTPHEADER => [
        'x-api-key: ' . $apiKey,
        'Accept: application/json',
    ],
]);

$body = curl_exec($ch);
$status = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
$err = curl_error($ch);
curl_close($ch);

if ($body === false) {
    fwrite(STDERR, "curl error: {$err}\n");
    exit(1);
}
if ($status < 200 || $status >= 300) {
    fwrite(STDERR, "HTTP {$status}: {$body}\n");
    exit(1);
}

$data = json_decode($body, true);
if (!is_array($data) || !isset($data['result'])) {
    fwrite(STDERR, "Unexpected response shape\n");
    exit(1);
}

$result = $data['result'];
// Prefer formatted_address for display; dump full object for debugging.
echo ($result['formatted_address'] ?? '(no formatted_address)') . PHP_EOL;
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES) . PHP_EOL;
